tobac/utils/decorators.py

```diff
@@ -5,344 +5,369 @@
-                        for output_item in output
+                        for arg in args
                     ]
                 )
-            elif type(output) == xarray.DataArray:
-                output = xarray.DataArray.to_iris(output)
-            # if output is neither tuple nor an xr.DataArray
+                kwargs_new = dict(
+                    zip(
+                        kwargs.keys(),
+                        [
+                            xarray.DataArray.from_iris(arg)
+                            if type(arg) == iris.cube.Cube
+                            else arg
+                            for arg in kwargs.values()
+                        ],
+                    )
+                )
+                # print(args)
+                # print(kwargs)
+                output = func(*args, **kwargs_new)
+                if type(output) == tuple:
+                    output = tuple(
+                        [
+                            xarray.DataArray.to_iris(output_item)
+                            if type(output_item) == xarray.DataArray
+                            else output_item
+                            for output_item in output
+                        ]
+                    )
+                elif type(output) == xarray.DataArray:
+                    output = xarray.DataArray.to_iris(output)
+                # if output is neither tuple nor an xr.DataArray
+                else:
+                    output = func(*args, **kwargs)
+
             else:
                 output = func(*args, **kwargs)
+            return output
 
-        else:
-            output = func(*args, **kwargs)
-        return output
-
-    return wrapper
+        return wrapper
 
+    return iris_to_xarray_i
 
-def xarray_to_iris(func):
-    """Decorator that converts all input of a function that is in the form of
-    xarray DataArrays into Iris cubes and converts all outputs with type
-    Iris cubes back into xarray DataArrays.
 
-    Parameters
-    ----------
-    func : function
-        Function to be decorated.
+def xarray_to_iris():
+    def xarray_to_iris_i(func):
+        """Decorator that converts all input of a function that is in the form of
+        xarray DataArrays into Iris cubes and converts all outputs with type
+        Iris cubes back into xarray DataArrays.
 
-    Returns
-    -------
-    wrapper : function
-        Function including decorator.
+        Parameters
+        ----------
+        func : function
+            Function to be decorated.
 
-    Examples
-    --------
-    >>> segmentation_xarray = xarray_to_iris(segmentation)
+        Returns
+        -------
+        wrapper : function
+            Function including decorator.
 
-    This line creates a new function that can process xarray fields and
-    also outputs fields in xarray format, but otherwise works just like
-    the original function:
+        Examples
+        --------
+        >>> segmentation_xarray = xarray_to_iris(segmentation)
```

Comment:
This example should be updated to the new structure of the decorators. And maybe it makes sense to move the docstrings to the outer functions. I don't really know how docstrings of inner functions are handled in the documentation.

Reply:
I've resolved the examples. I have so far not touched the decorator docstrings. We don't render these in the readthedocs, so exact placement isn't critical. I am inclined to leave them where they are for now to save a refactor.